### DISCUSSION TITLE
High CPU usage by AttackMapServer.py

### DISCUSSION THREAD

**Discussion by ChristophRh** [telekom-security/tpotce#1241] [Q&A] [answered]
Today I recognized a very high CPU usage by the AttackMapServer.py which was not the case a week ago.
Anyone else having issues?

There was an update of the [geoip_attack_map repository](https://github.com/t3chn0m4g3/geoip-attack-map) a few days ago. Could this update, which changed important code parts, be the reason for the high CPU usage or is it just an issue on my side?

Thanks for your feedback!

**Comment by t3chn0m4g3** [comment, marked as answer]
Can you give more info on your setup, hardware, average events per day, connected clients, etc.?

It should be noted however that the AttackMapServer works differently now. Except for MapBox all code is served locally, with aiohttp, aioredis and asyncio to ensure better performance. Also I activated gzip and brotli support in NGINX and deferred the loading of JS code which will result in lots of parallel connections on page load.

Should you spot errors in the code, the setup or while debugging please share.

**Reply by ChristophRh** [reply]
I have TPOT running on an AWS instance t2.large with 8GB RAM and 2 cores. 
There are about 45000 attack events per day.
Most of the time there is 0 or just 1 client connected. If the client for purposes of demonstration is connected it shows the attack map and 5 Kibana dashboards that are updating every 30 seconds.
So far I had no issues with this setup (except for ConPot which sometimes stucks at high CPU usage, but that is an other problem). Overall performance was good.

Now without any client connected the AttackMapServer.py is using up to 100% of 1 core.
When I deactivate the attack map in tpot.yml settings the CPU usage is normal.

Another issue now is that the geolocation of my instance is wrong. It is somewhere in Seattle (US) instead of Frankfurt am Main (GER). I have to mention that it was the same in Kibana live attack map, but in this Attack map it was correct before.

If it is like it is with the new attack map, is there a way to use the old attack map and preventing TPOT from updating it?
I did not manage to do that so far.

**Reply by t3chn0m4g3** [reply]
Thanks for takting the time to report. I can reproduce the issue with high CPU usage and have an idea what the reason is and how to fix it. I opened the issue [here](https://github.com/t3chn0m4g3/geoip-attack-map/issues/3) for tracking.

**Reply by ChristophRh** [reply]
Sounds great, thank you very much!

**Reply by t3chn0m4g3** [reply]
The high CPU usage is fixed now. If you want to test right away, stop T-Pot and pull the corresponding image.
```
systemctl stop tpot
docker pull dtagdevsec/map:2204
systemctl start tpot
```

Regarding the geoip info. The new map no longer uses the locally derived geoIP info (based on where the map was started [from](https://github.com/telekom-security/tpotce/blob/master/docker/elk/map/dist/entrypoint.sh)), but the local geoIP of each individual T-Pot (`geoip_ext.ip`).

**Reply by ChristophRh** [reply]
Works perfectly, thanks a lot for the quick fix!

Any idea where these wrong values of my `geoip_ext.ip` originate from?
Should I open a new topic for this issue?

**Reply by t3chn0m4g3** [reply]
Great, thanks for the swift feedback!

> Any idea where these wrong values of my geoip_ext.ip originate from?

Please check the Kibana Dashboard > Discover and search for some events. Enlarge the events and check its fields for `geoip_ext.ip` and compare it to the `t-pot_ip_ext` field. Typically these offer the same IP (identical to `/etc/issue`).

Since it is identical to the one from Kibana, as you mentioned earlier, it is the correct behaviour and necessary to ensure dynamic mapping of T-Pots in distributed scenarios.

**Reply by ChristophRh** [reply]
`geoip_ext.ip` and `t-pot_ip_ext` are identical and the same as the public IP listet in `/etc/issue`.
But still `geoip_ext.latitude` and `geoip_ext.longitude` are pointing to Seattle which makes no sense. If I lookup the geolocation of my IP address it is in Germany and that should be correct since I'm using an AWS instance in Frankfurt.

I know that this issue is not related to your attack map anymore, but I thought you could give me a quick hint where these wrong values come from. But I can also start a new discussion for this topic.

**Reply by t3chn0m4g3** [reply]
This is **geopoint** related. For the old map the local geopoint was determined from ipinfo (as seen [here](https://github.com/telekom-security/tpotce/blob/master/bin/updateip.sh)) while the new one uses geo point from Logstash (which uses [GeoLite](https://github.com/telekom-security/tpotce/blob/516bec1deb59b4a0e0abb889e0458b9752c55bd0/docker/elk/logstash/dist/logstash.conf#L655-L668)); ipinfo had too many disadvantages (i.e. only one T-Pot supported) and will be removed entirely. As long as Logstash does not receive a GeoLite DB update the info on your end will be not the same as before. You can however use [volumes](https://github.com/telekom-security/tpotce/wiki/Reconfigure-logstash.conf#5-adjust-tpotyml-by-adding-docker-volume-for-logstashconf) and exchange the GeoLite database on your own.

**Reply by ChristophRh** [reply]
I managed to do that :)
Thanks a lot for your help!

**Comment by bob3rocks** [comment]
Can you tell me what the fix was for this? I've got the same issue, had to disable AttackMapServer in the tpot config due to sustained 90%+ CPU. I installed T-Pot 2.3 two weeks ago, and it was working great for the first few days.  Thanks

**Reply by t3chn0m4g3** [reply]
The new image should have downloaded already, if not just follow the steps outlined above. The new image solves the issue.

**Comment by bob3rocks** [comment]
Gotcha! Thanks, I will try that

On Sun, Jan 15, 2023 at 1:17 PM Marco Ochse ***@***.***>
wrote:

> The new image should have downloaded already, if not just follow the steps
> outlined above. The new image solves the issue.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/telekom-security/tpotce/discussions/1241#discussioncomment-4691970>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFPYFARME42ZLAVV2HENCDTWSRSP3ANCNFSM6AAAAAATVR2ASM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>


-- 

Bob Wachunas
Breakaway Funk <http://breakawayfunk.com/>
707-536-1182

**Comment by bob3rocks** [comment]
Looks good now!

Thanks

On Sun, Jan 15, 2023 at 1:17 PM Marco Ochse ***@***.***>
wrote:

> The new image should have downloaded already, if not just follow the steps
> outlined above. The new image solves the issue.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/telekom-security/tpotce/discussions/1241#discussioncomment-4691970>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFPYFARME42ZLAVV2HENCDTWSRSP3ANCNFSM6AAAAAATVR2ASM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>


-- 

Bob Wachunas
Breakaway Funk <http://breakawayfunk.com/>
707-536-1182

**Comment by t3chn0m4g3** [comment]
Deployed [1.1.0](https://github.com/t3chn0m4g3/t-pot-attack-map/releases/tag/1.1.0) yesterday. Let me know what you think.

**Reply by ChristophRh** [reply]
It failed to load resource `PS.png`, which should be the flag of Palestine I guess.
And today the map did not load any attacks, but after a restart of tpot it works again. I will keep an eye on that.

Everything else looks great, thank you!
But is there a way to prevent the attack map from updating automatically so that I can choose when to update?

**Reply by t3chn0m4g3** [reply]
Thanks, I will be updating all flags with Flagpack shortly.

Check the [Readme](https://github.com/telekom-security/tpotce#updates) for details on updates and daily reboot.

**Reply by ChristophRh** [reply]
The Readme does not contain enough details for me. I just want to exclude the map from automatic updates. 
Can I do that by removing the line `image: "dtagdevsec/map:2204"` in `/opt/tpot/etc/tpot.yml` or does that break something.
I'm not experienced with docker.

**Reply by t3chn0m4g3** [reply]
It is either all or nothing (`/etc/crontab`), removing the image line will result in an error and T-Pot will not start.

**Reply by ChristophRh** [reply]
okay
thank you very much for your help!
